pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -673,11 +673,21 @@ func (r *ReconcileBuildRun) createTaskRun(ctx context.Context, build *buildv1alp
 	return generatedTaskRun, nil
 }
 
+// updateBuildRunErrorStatus updates buildRun status fields
 func (r *ReconcileBuildRun) updateBuildRunErrorStatus(ctx context.Context, buildRun *buildv1alpha1.BuildRun, errorMessage string) error {
+	// these two fields are deprecated and will be removed soon
 	buildRun.Status.Succeeded = corev1.ConditionFalse
 	buildRun.Status.Reason = errorMessage
+
 	now := metav1.Now()
 	buildRun.Status.CompletionTime = &now
+	buildRun.Status.SetCondition(&buildv1alpha1.Condition{
+		LastTransitionTime: now,
+		Type:               buildv1alpha1.Succeeded,
+		Status:             corev1.ConditionFalse,
+		Reason:             "Failed",
```

Comment:
I was just looking through the root causes when we go into that function. They imo can be categorized into two groups:

1. User-originated, for example if the build does not exist or is not registered or the service account does not exist
2. System-caused, for example if we fail to create a TaskRun or to retrieve the build strategy (which should be there because the build is registered)

For the second group, I think it is okay to use `Failed` as reason. For the first group, I think we should have specific reasons that allow client applications to identify this situation and to present a localized error message to the end user.

Reply:
FWIW We follow an analogous approach for "system-caused" failures with OpenShift Builds.  There, system-caused generally means the build pod could not be constructed.  It has a unique error code compared to the "user errors"

so +1

Reply:
@SaschaSchwarze0 @gabemontero good feedback. The categorization is accurate. To provide more clarity on categorizations only for `FALSE` Status, we currently have the following:

_Note:_ Please see the current [docs](https://github.com/shipwright-io/build/blob/master/docs/buildrun.md#understanding-the-state-of-a-buildrun)

| Status | Reason | CompletionTime is set | Description |
| --- | --- | --- | --- |
| False    | Failed | Yes | The BuildRun failed in one of the steps. |
| False    | BuildRunTimeout | Yes | The BuildRun timed out. |

from this discussion, I think the enhancement will be to extend more flavors of `False` Status, as follows:

| Status | Reason | CompletionTime is set | Description |
| --- | --- | --- | --- |
| False    | Failed | Yes | The BuildRun failed in one of the steps. |
| False    | BuildRunTimeout | Yes | The BuildRun timed out. |
| False    | BuildRegistrationFailed | Yes | The referenced Build is not register correctly. |
| False    | TaskRun is missing | Yes | The related TaskRun is missing. |
| False    | UnknownStrategy | Yes | The referenced Strategy is unknown. |

I´m adding three more custom `Reason`´s. We currently do nine calls to `updateBuildRunErrorStatus` , and from this nine, only three have customized errors, the other six are related to sys calls (_e.g. GET, CREATE, etc)

Because of the amount of work we need for introducing the above three new flavors of a `False` Status, I suggest to open a new issue for this enhancement. The current PR needs to get merged, otherwise we are not populating our `Status.Conditions` on errors.

Reply:
Is fine with me. The lack of the missing condition also has a negative impact on e2e tests because they only check the status on the condition and therefore hang in the `Eventually` loop until this times out (I observed that yesterday in our downstream test when we had the network issue). As stated in [How should we treat a BuildRun? #558](https://github.com/shipwright-io/build/issues/558), `updateBuildRunErrorStatus` is playing a big role there as well. So, once the merge conflicts are resolved we can go that intermediate step.

Reply:
Conflicts have been resolved.